File: frontend/src/rpc/client/akun.js
import TAkunService from '@/rpc/gen/TAkunService';
import { TBaseClient } from '@/rpc/client/base';
import { TUserRole } from '@/rpc/gen/auth_types';


class TAkunServiceClient extends TBaseClient{
	constructor(stores=null){
		super(stores, TAkunService, '/api/akun/akun');
	}

	async fetch_akun(query){
		this.stores.helper.auth.requireRole(TUserRole.ADMIN_AKUN);
		return await this.client.fetch_akun(this.stores.auth.authToken, query);
	}
	async register_akun(form){
		this.stores.helper.auth.requireRole(TUserRole.ADMIN_AKUN);
		return await this.client.register_akun(this.stores.auth.authToken, form);
	}

	async set_role(user_id, new_role){
		this.stores.helper.auth.requireRole(TUserRole.ADMIN_AKUN);
		await this.client.set_role(this.stores.auth.authToken, user_id, new_role);
		if (user_id == this.stores.auth.user.id){
			//await this.stores.auth.setUserRole(new_role);
			await this.stores.client.auth.logout();
		}
	}

	async set_email(user_id, new_email){
		this.stores.helper.auth.requireRole(TUserRole.ADMIN_AKUN);
		await this.client.set_email(this.stores.auth.authToken, user_id, new_email);
	}
	async set_password(user_id, new_password){
		this.stores.helper.auth.requireRole(TUserRole.ADMIN_AKUN);
		await this.client.set_password(this.stores.auth.authToken, user_id, new_password);
		if (user_id == this.stores.auth.user.id){
			await this.stores.client.auth.login(this.stores.auth.user.username, new_password);
		}
	}
	async set_enabled(user_id, new_enabled){
		this.stores.helper.auth.requireRole(TUserRole.ADMIN_AKUN);
		await this.client.set_enabled(this.stores.auth.authToken, user_id, new_enabled);
		if (user_id == this.stores.auth.user.id && !new_enabled){
			await this.stores.client.auth.logout();
		}
	}
	

}

const akunClient = new TAkunServiceClient();

export { TAkunServiceClient, akunClient }
export default akunClient

File: backend/rpc/handler/akun/akun.py
from rpc.gen.akun.akun import TAkunService
from models import get_model
from rpc.gen.akun.auth.ttypes import TAuthError, TAuthErrorCode, TUserRole
from rpc.gen.akun.user.ttypes import TUser, TUserError, TUserErrorCode
from converter.user import DBUser_TUser

class TAkunServiceHandler(TAkunService.Iface):
    def __init__(self):
        self.auth_model = get_model('auth')
        self.user_model = get_model('user')

    def fetch_akun(self, auth_token, query):
        auth_payload = self.auth_model.require_role(auth_token, TUserRole.ADMIN_AKUN)
        db_users = self.user_model.fetch_users(query)
        return [DBUser_TUser(u) for u in db_users]

    def register_akun(self, auth_token, form):
        auth_payload = self.auth_model.require_role(auth_token, TUserRole.ADMIN_AKUN)
        db_user = self.user_model.register_user(auth_payload['role'], form)
        return DBUser_TUser(db_user)

    def set_role(self, auth_token, user_id, new_role):
        auth_payload = self.auth_model.require_role(auth_token, TUserRole.ADMIN_AKUN)
        self.user_model.set_role(auth_payload['role'], user_id, new_role)


    def set_email(self, auth_token, user_id, new_email):
        auth_payload = self.auth_model.require_role(auth_token, TUserRole.ADMIN_AKUN)
        self.user_model.set_email(auth_payload['role'], user_id, new_email)

    def set_password(self, auth_token, user_id, new_password):
        auth_payload = self.auth_model.require_role(auth_token, TUserRole.ADMIN_AKUN)
        self.user_model.set_password(auth_payload['role'], user_id, new_password)

    def set_enabled(self, auth_token, user_id, new_enabled):
        auth_payload = self.auth_model.require_role(auth_token, TUserRole.ADMIN_AKUN)
        self.user_model.set_enabled(auth_payload['role'], user_id, new_enabled)
